### PR TITLE
Report cancellation reason to GA/S3

### PR DESCRIPTION
**Pull request by Naatan**
https://www.pivotaltracker.com/story/show/174811635

### CODE REVIEW

File: installers/msi-language/CustomUserExitDialog.wxs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+            <Dialog Id="CustomUserExitDialog" Width="370" Height="270" Title="!(loc.UserExit_Title)">
+                <Control Id="Finish" Type="PushButton" X="236" Y="243" Width="56" Height="17" Default="yes" Cancel="yes" Text="!(loc.WixUIFinish)">
+                    <Publish Event="EndDialog" Value="Exit">1</Publish>
+                </Control>
+                <Control Id="Cancel" Type="PushButton" X="304" Y="243" Width="56" Height="17" Disabled="yes" Text="!(loc.WixUICancel)" />
+                <Control Id="Bitmap" Type="Bitmap" X="0" Y="0" Width="370" Height="234" TabSkip="no" Text="!(loc.UserExitBitmap)" />
+                <Control Id="Back" Type="PushButton" X="180" Y="243" Width="56" Height="17" Disabled="yes" Text="!(loc.WixUIBack)" />
+                <Control Id="BottomLine" Type="Line" X="0" Y="234" Width="370" Height="0" />
+                <Control Id="Title" Type="Text" X="135" Y="20" Width="220" Height="60" Transparent="yes" NoPrefix="yes" Text="!(loc.UserExitTitle)" />
+                <Control Id="Description" Type="Text" X="135" Y="80" Width="220" Height="80" Transparent="yes" NoPrefix="yes" Text="!(loc.UserExitDescription1) !(loc.UserExitDescription2)" />
```
Comment (MDrakos) [inline installers/msi-language/CustomUserExitDialog.wxs:25]: I'm not seeing a diff for the localization file. Are `UserExitDescription1` and `UserExitDescription2` built-in?

File: installers/msi-language/Product.p.wxs [hunk 2]
```diff
@@ -213,11 +222,16 @@
 		<CustomAction
 			Id="SetStateDeployPropertyValues"
 			Property="StateDeploy"
-			Value="INSTALLDIR=[INSTALLDIR];PROJECT_OWNER_AND_NAME=[PROJECT_OWNER_AND_NAME];IS_MODIFY=[IS_MODIFY];STATE_TOOL_INSTALLED=[STATE_TOOL_INSTALLED];STATE_TOOL_PATH=[STATE_TOOL_PATH];AS_USERNAME=[AS_USERNAME];AS_PASSWORD=[AS_PASSWORD];AS_TOTP=[AS_TOTP];UI_LEVEL=[UILevel];MSI_VERSION=[MSI_VERSION];PRODUCT_VERSION=[ProductVersion];MsiLogFileLocation=[MsiLogFileLocation];PRIVACY_ACCEPTED=[PRIVACY_ACCEPTED];ERROR=[ERROR];ERROR_MESSAGE=[ERROR_MESSAGE];PRODUCT_NAME=[ProductName];USERSID=[UserSID];COMMIT_ID=[CommitID];INSTALL_MODE=[INSTALL_MODE]"
+			Value="INSTALLDIR=[INSTALLDIR];PROJECT_OWNER_AND_NAME=[PROJECT_OWNER_AND_NAME];IS_MODIFY=[IS_MODIFY];STATE_TOOL_INSTALLED=[STATE_TOOL_INSTALLED];STATE_TOOL_PATH=[STATE_TOOL_PATH];AS_USERNAME=[AS_USERNAME];AS_PASSWORD=[AS_PASSWORD];AS_TOTP=[AS_TOTP];UI_LEVEL=[UILevel];MSI_VERSION=[MSI_VERSION];PRODUCT_VERSION=[ProductVersion];MsiLogFileLocation=[MsiLogFileLocation];PRIVACY_ACCEPTED=[PRIVACY_ACCEPTED];ERROR=[ERROR];ERROR_MESSAGE=[ERROR_MESSAGE];PRODUCT_NAME=[ProductName];USERSID=[UserSID];COMMIT_ID=[CommitID];INSTALL_MODE=[INSTALL_MODE];LAST_DIALOG=[LAST_DIALOG]"
```
Comment (MDrakos) [inline installers/msi-language/Product.p.wxs:225]: I don't think it's necessary to pass `LAST_DIALOG` into any of the deferred custom actions, everything you're working with here is in an `immediate` context. (ie. you're not using `session.CustomActionData["LAST_DIALOG"]` anywhere)

File: installers/msi-language/Product.p.wxs [hunk 3]
```diff
@@ -213,11 +222,16 @@
...
-		<CustomAction Id="SetRollbackPropertyValues" Property="Rollback" Value="STATE_TOOL_INSTALLED=[STATE_TOOL_INSTALLED];INSTALLDIR=[INSTALLDIR];MSI_VERSION=[MSI_VERSION];MsiLogFileLocation=[MsiLogFileLocation];PRIVACY_ACCEPTED=[PRIVACY_ACCEPTED]" />
+		<CustomAction Id="SetPresetInfo" Property="Preset" Value="REL_NOTES=[REL_NOTES];INSTALLDIR=[INSTALLDIR];APP_START_MENU_PATH=[ApplicationProgramsFolder];PRESET=[PRESET];MSI_VERSION=[MSI_VERSION];MsiLogFileLocation=[MsiLogFileLocation];PRIVACY_ACCEPTED=[PRIVACY_ACCEPTED];LAST_DIALOG=[LAST_DIALOG];INSTALL_MODE=[INSTALL_MODE]" />
+		<CustomAction Id="SetRollbackPropertyValues" Property="Rollback" Value="STATE_TOOL_INSTALLED=[STATE_TOOL_INSTALLED];INSTALLDIR=[INSTALLDIR];MSI_VERSION=[MSI_VERSION];MsiLogFileLocation=[MsiLogFileLocation];PRIVACY_ACCEPTED=[PRIVACY_ACCEPTED];LAST_DIALOG=[LAST_DIALOG];INSTALL_MODE=[INSTALL_MODE]" />
+		<CustomAction
+			Id="SetCancelPropertyValues"
+			Property="CustomUserExit"
+			Value="INSTALL_MODE=[INSTALL_MODE];PRODUCT_VERSION=[ProductVersion];MsiLogFileLocation=[MsiLogFileLocation];LAST_DIALOG=[LAST_DIALOG]"
+		/>
```
Comment (MDrakos) [inline installers/msi-language/Product.p.wxs:234]: This custom action is created but never scheduled. You would need this if `CustomUserExit` was deferred but it's not

**Review by MDrakos — CHANGES_REQUESTED** [review]

File: installers/msi-language/CustomUserExitDialog.wxs (hunk 1 above)
Comment (Naatan) [reply installers/msi-language/CustomUserExitDialog.wxs:25]: Yeah this is copy/pasted from the UserExit.wxs file. I tested it and it seems to be showing the right values.

**Pull request merged**